Mathlib/CategoryTheory/Topos/Basic.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2024 Charlie Conneen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Charlie Conneen
+-/
+import Mathlib.CategoryTheory.ChosenFiniteProducts
+import Mathlib.CategoryTheory.Topos.Power
+
+
+/-!
+# Topoi
+
+We follow Mac Lane and Moerdijk in defining a topos as a category
+with a terminal object, pullbacks, a subobject classifier, and
+power objects.
+
+In this file, a "predicate" refers to any morphism into `Ω C`.
+
+## Main definitions
+
+* `CategoryTheory.Topos C` is a typeclass asserting that `C` is a topos.
+
+* The namespace `CategoryTheory.Topos.Predicate` provides an API for various
+  useful predicates, such as `Predicate.eq` as the equality predicate
+  `B ⨯ B ⟶ Ω C`, and `Predicate.isSingleton` which is the classifier of the
+  singleton map `singleton B : B ⟶ Pow B`.
+
+## Main results
+
+* `singleton B : B ⟶ Pow B` is a monomorphism. This is `singletonMono`.
+
+## Notation
+
+* If `φ` is a predicate `X ⟶ Ω C`, `⌈φ⌉` is shorthand for `Name φ`,
+  which is the global element `⊤_ C ⟶ Pow X` associated to `φ`.
+
+## References
+
+* [S. MacLane and I. Moerdijk, *Sheaves in Geometry and Logic*][MLM92]
+
+-/
+
+namespace CategoryTheory
+
+open Category Limits Classifier Power
+
+universe u v
+
+variable (C : Type u) [Category.{v} C]
+
+/-- A category is a topos if it has a terminal object,
+all pullbacks, a subobject classifier, and power objects.
+-/
+class Topos where
+  /-- A topos has a terminal object. -/
+  [has_terminal : HasTerminal C]
+  /-- A topos has pullbacks. -/
+  [has_pullbacks : HasPullbacks C]
+  /-- A topos has a subobject classifier. -/
+  [subobject_classifier : HasClassifier C]
+  /-- A topos has power objects. -/
+  [has_power_objects : HasPowers C]
+
+attribute [instance] Topos.has_terminal Topos.has_pullbacks
+                     Topos.subobject_classifier Topos.has_power_objects
+
+variable [Topos C] {C}
+
+namespace Topos
+
+/-- A topos has (chosen) finite products. -/
+noncomputable instance chosenFiniteProducts : ChosenFiniteProducts C :=
```

Comment:
This should be an assumption rather than an instance, because `ChosenFiniteProducts` contains data. Otherwise, when we want to apply this to sheaf categories, we shall get multiple instances.